pkg/btree/btree.go

```diff
@@ -48,3 +51,173 @@ func (t *BTree) readNode(offset uint64) (*BTreeNode, error) {
+	}
+
+	parent, parentOffset := root, rootOffset.Offset
+	for !parent.Leaf() {
+		index, found := slices.BinarySearchFunc(parent.Ids, key, pointer.CompareReferencedIds)
+
+		if found {
+			panic("cannot insert duplicate key")
+		}
+
+		loffset := parent.Offsets[index]
+		child, err := t.readNode(loffset)
+		if err != nil {
+			return err
+		}
+
+		if int(child.Size()) > t.PageFile.PageSize() {
+			// split node here
+			mid := len(child.Ids) / 2
+			midKey := child.Ids[mid]
+
+			rightChild := &BTreeNode{Width: t.Width, VectorDim: t.VectorDim}
+			if !child.Leaf() {
+				rightChild.Offsets = child.Offsets[mid+1:]
+				child.Offsets = child.Offsets[:mid]
+			}
+			rightChild.Vectors = child.Vectors[mid+1:]
+			rightChild.Ids = child.Ids[mid+1:]
+
+			rbuf, err := rightChild.MarshalBinary()
+			if err != nil {
+				return err
+			}
+			roffset, err := t.PageFile.NewPage(rbuf)
+			if err != nil {
+				return err
+			}
+
+			// shrink left child (child)
+			child.Ids = child.Ids[:mid]
+			child.Vectors = child.Vectors[:mid]
+			if _, err := t.PageFile.Seek(int64(loffset), io.SeekStart); err != nil {
+				return err
+			}
+
+			if _, err := child.WriteTo(t.PageFile); err != nil {
+				return err
+			}
+
+			// update parent to include new key and store left right offsets
+			if index == len(parent.Ids) {
+				parent.Ids = append(parent.Ids, midKey)
+			} else {
+				parent.Ids = append(parent.Ids[:index+1], parent.Ids[index:]...)
+				parent.Ids[index] = midKey
+			}
+
+			parent.Offsets = append(parent.Offsets[:index+2], parent.Offsets[:index+1]...)
+			parent.Offsets[index+1] = uint64(roffset)
+			if _, err := t.PageFile.Seek(int64(parentOffset), io.SeekStart); err != nil {
+				return err
+			}
+			if _, err := parent.WriteTo(t.PageFile); err != nil {
+				return err
+			}
+
+			cmpRes := pointer.CompareReferencedIds(midKey, key)
+			if cmpRes == 0 {
+				panic("midKey and key should not be the same post split detection")
+			}
+
+			if cmpRes == 1 {
+				parent, parentOffset = child, loffset
+			} else {
+				parent, parentOffset = rightChild, uint64(roffset)
+			}
+		} else {
+			if _, err := t.PageFile.Seek(int64(parentOffset), io.SeekStart); err != nil {
+				return err
+			}
+			if _, err := parent.WriteTo(t.PageFile); err != nil {
+				return err
+			}
+			parent, parentOffset = child, loffset
+		}
+	}
+
+	index, found := slices.BinarySearchFunc(parent.Ids, key, pointer.CompareReferencedIds)
+	if found {
+		panic("cannot insert duplicate key")
+	}
+
+	if index == len(parent.Ids) {
+		parent.Ids = append(parent.Ids, key)
+		parent.Vectors = append(parent.Vectors, vector)
+	} else {
+		parent.Ids = append(parent.Ids[:index+1], parent.Ids[index:]...)
+		parent.Ids[index] = key
+		parent.Vectors = append(parent.Vectors[:index+1], parent.Vectors[index:]...)
+		parent.Vectors[index] = vector
+	}
+
+	// this should be the leaf node
+	if int(parent.Size()) > t.PageFile.PageSize() {
+		panic("too fat!")
+	}
```

Comment:
When we reach the point where we finally serialize the `leaf` node post top-down traversal, that `leaf` size can exceed a page. I was going to split, but is creating an intermediate node with only one element (the mid key) be sufficient?

Reply:
Hmm I think splitting while traversing downwards won't work because of this problem. Adding an intermediate node isn't great because there is a degenerate sequence of inserts that will always trigger an intermediate node insertion leading to bad asymptotic performance.

Specifically, right now you're splitting the tree as you're traversing down to the leaf node but instead it might be easier to insert it first and then propagate the split back _up_ the tree. This would fix the issue as you wouldn't need to create an intermediate node.

Reply:
Sounds good